### PR TITLE
feat: Ore localization of a commutative algebra is an algebra.

### PR DESCRIPTION
**Pull request by kbuzzard**
If A is a commutative R-algebra and S is a submonoid of R then A[1/S] is an R[1/S]-algebra. The structure morphism R[1/S] -> A[1/S] is defined using a lifting property, and in particular if `A = R` it turns out to be not defeq to `algebra.id R`, so I've not made it an instance. In a proof I'm about to embark on, I'll probably be switching it on locally.

To avoid diamonds, I've had to remove the irreducibility tag from `OreLocalization.smul` , `OreLocalization.one` and `OreLocalization.npow`. This change seems to make the code base cleaner (I removed some `unseal`s and added no `seal`s), so I removed all the other ones as well. This has resulted in a slowdown in one file upstream; I isolate the problem in #16746 . 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/OreLocalization/Algebra.lean [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+protected theorem one_def' : (1 : A[S⁻¹]) = 1 /ₒ 1 := rfl
+
+end One
+
+section CommMonoid
+
+variable {R A : Type*} [CommMonoid R] [CommMonoid A] [MulAction R A] [IsScalarTower R A A]
+  {S : Submonoid R}
+
+instance commMonoid' : CommMonoid (A[S⁻¹]) where
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/OreLocalization/Algebra.lean:88]: I'm thinking instances should have the `inst` prefix

File: Mathlib/RingTheory/OreLocalization/Algebra.lean [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      one_mul, zero_oreDiv, zero_oreDiv]
+
+end CommSemiring
+
+section Algebra
+
+variable (R A : Type*) [cR : CommSemiring R] [cA : CommSemiring A] [cRA : Algebra R A]
+    (S : Submonoid R)
+
+def instAlgebra' : Algebra (R[S⁻¹]) (A[S⁻¹]) where
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/OreLocalization/Algebra.lean:169]: while defs shouldn't

File: Mathlib/RingTheory/OreLocalization/Algebra.lean [hunk 3]
```diff
@@ -0,0 +1,210 @@
+import Mathlib.RingTheory.OreLocalization.Ring
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/OreLocalization/Algebra.lean:1]: You're going to want to add a copyright header

File: Mathlib/RingTheory/OreLocalization/Algebra.lean (hunk 2 above)
Comment (kbuzzard) [reply Mathlib/RingTheory/OreLocalization/Algebra.lean:169]: It was an instance until Eric pointed out diamond possibilities and then I found one. It's the example at the bottom of the file. Can I make this defeq somehow?

**Comment by kbuzzard** [comment]
!bench

**Comment by kbuzzard** [comment]
I don't need this code any more, I only need these ideas in the commutative case and I'd misunderstood how to do this.

**Pull request closed without merging**